### ISSUE TITLE
Unable to proxy Socks5 in beta

### ISSUE THREAD

**Issue by bbhxwl** [justcoding121/Titanium-Web-Proxy#721]
I ran titanium.web.proxy.examples.basic, released the comments of the following code, tested the native 127.0.0.1 and port 1080, unable to run normally?

 

var socksEndPoint = new SocksProxyEndPoint(IPAddress.Any, 1080, true)
            {
                // Generic Certificate hostname to use
                // When SNI is disabled by client
                GenericCertificateName = "google.com"
            };

           proxyServer.AddEndPoint(socksEndPoint);

**Comment by honfika** [comment]
What is the problem?

**Comment by bbhxwl** [comment]
> What is the problem?

Is this the key code of Socks5? Did I test or couldn't support Socks5?

**Comment by honfika** [comment]
Yes, that is the code.

**Comment by bbhxwl** [comment]
After I enable it, I can't connect to Socks5 properly. Is there a bug?

**Comment by bbhxwl** [comment]
There is a problem. UDP function cannot be proxy

**Comment by bbhxwl** [comment]
![QQ图片20191227193120](https://user-images.githubusercontent.com/26525599/71515671-7db42f00-28df-11ea-9278-d8be84c57fd6.png)

**Comment by honfika** [comment]
Yes, this is TCP only

**Comment by bbhxwl** [comment]
Will UDP compatible forwarding be considered in the future?

**Comment by honfika** [comment]
Could you please translate your screnshot? I can't speak chinese, and can1t copy the characters from an image:)

**Comment by honfika** [comment]
How do you use UDP Socks fro a browser? Or what is the source of your HTTP requests?

**Comment by bbhxwl** [comment]
The text of the picture is probably that the test TCP passed and UDP failed. If you want to be a proxy server, you can not only speed up HTTP and UDP, but also speed up the game. If it is compatible, it will be a more powerful project.

**Comment by honfika** [comment]
But TWP is a HTTP proxy... it accepts only HTTP requests in SOCKS protocol.... there are many simple socks proxy on the internet.

**Comment by bbhxwl** [comment]
Well, I want to integrate.

**Comment by bbhxwl** [comment]
Can't you make something like a plug-in? Support for UDP will not affect the TCP function of Socks5.

**Comment by bbhxwl** [comment]
> But TWP is a HTTP proxy... it accepts only HTTP requests in SOCKS protocol.... there are many simple socks proxy on the internet.

Does the official version support Socks5 authentication? Hope to consider adding a UDP forwarding.

**Comment by honfika** [comment]
Yes, it supports authentication, you have to set the ProxyBasicAuthenticateFunc property to authorize the connection.

**Comment by honfika** [comment]
However i did not fould any browser which supprots SOCKS5 authentication. It is possible only with a Firefox addon, FoxyProxy.

**Comment by bbhxwl** [comment]
I want to know HTTP authentication and sock5 authentication. In addition to limiting the account and password, I can get the IP of the client. I want to do the IP restriction

**Comment by bbhxwl** [comment]
If I don't want to intercept HTTPS and forward data directly, can I not write the onbeforetunnelconnectrequest event? Is the default e.decryptssl = false?

**Comment by honfika** [comment]
The default is "true".

If you don't want to decrypt any traffic, i suggest you to disable decryption globally in the endpoint constructor.

Currently the client infomration is not available for SOCKS authentication (the SessionEventArgsBase parameter is null)

**Comment by bbhxwl** [comment]
explicitEndPoint.DecryptSsl     is  read-only
If I don't want to intercept any data and just want to be an HTTP proxy, I can add no events, right? Just add an explicitproxyendpoint.

**Comment by honfika** [comment]
Set the **constructor parameter** of the endpoint object.

**Comment by bbhxwl** [comment]
Sock5 authentication cannot be restricted in proxybasicauthenticatefunc, right?



proxyServer.ProxyBasicAuthenticateFunc = async (args, userName, password) =>
            {

                return true;
            };


 var socksEndPoint = new SocksProxyEndPoint(IPAddress.Any, 1080,false)
            {
                // Generic Certificate hostname to use
                // When SNI is disabled by client
                GenericCertificateName = "google.com"
            };
             
            proxyServer.AddEndPoint(socksEndPoint);

**Comment by honfika** [comment]
It can be restricted as you wrote.

**Comment by bbhxwl** [comment]
So you can limit sock5? Recommend the author's S5 client program sstap and C ා open source program netch. Is the author not going to support UDP forwarding?

**Comment by bbhxwl** [comment]
Is there a way that the user can not enter the password? I will judge the authorization according to the IP. If the user enters the account and password, I will judge with the account and password. It seems that after writing proxybasicauthenticatefunc, you must enter the account and password, right?

**Comment by honfika** [comment]
Yes, if you set the proxybasicauthenticatefunc the user must enter the username and password.

**Comment by bbhxwl** [comment]
I have written these codes that can be used normally. There are some parameters that I don't understand.

For example, I don't know proxyserver.enablewinauth

What is the function of this parameter? Is args.clientremoteendpoint.address.maptoipv4() in proxyserver.proxybasicauthenticatefunc the IP address of the connector? Is proxyendpoint the IP address of the connected (server)? Should clientlocalendpoint be the server's native IP?

Is proxyserver.forwardtoupstreamgateway important? What will happen if I cancel? Does enabling http2 have any effect on the load?

proxyServer = new ProxyServer();
                proxyServer.EnableHttp2 = true;
                proxyServer.ExceptionFunc = exception =>
                {

                };
                proxyServer.ForwardToUpstreamGateway = true;
proxyServer.BeforeRequest += onRequest;
                proxyServer.BeforeResponse += onResponse;
                proxyServer.EnableWinAuth = true;
                proxyServer.ProxyBasicAuthenticateFunc = async (args, userName, password) =>
                {
                   return true;
                 };
                explicitEndPoint = new ExplicitProxyEndPoint(IPAddress.Any, TcpPort, false);
                proxyServer.AddEndPoint(explicitEndPoint);
                proxyServer.Start();

**Comment by justcoding121** [comment]
Closing with the current SOCKS capability statement after the #860 auth work:

**Supported**
- SOCKS5 and SOCKS4 **TCP CONNECT**
- Opaque TCP relay after CONNECT (including non-HTTP binary traffic)
- SOCKS5 username/password auth (endpoint + global callbacks; see #860)

**Not supported**
- SOCKS **UDP ASSOCIATE**
- SOCKS **BIND**

No separate UDP roadmap issue is being opened at this time; UDP ASSOCIATE/BIND remain out of scope unless demand justifies a dedicated design later.

**Comment by justcoding121** [comment]
Closing: TCP CONNECT + opaque relay are supported; UDP ASSOCIATE/BIND are not. See previous comment.
